tests/Aql/LimitExecutorTest.cpp

```diff
@@ -116,137 +120,987 @@ TEST_F(LimitExecutorTest, rows_upstream_the_producer_doesnt_wait_limit_1_offset_
+  EXPECT_EQ(0, stats.getFullCount());
+  fullCount += stats.getFullCount();
+
+  std::tie(state, stats) = testee.produceRows(row);
+  ASSERT_EQ(ExecutionState::DONE, state);
+  ASSERT_TRUE(row.produced());
+  EXPECT_EQ(1, stats.getFullCount());
+  fullCount += stats.getFullCount();
+
+  ASSERT_EQ(4, fullCount);
 
   auto block = row.stealBlock();
+  EXPECT_EQ(3, block->size());
   AqlValue value = block->getValue(0, 0);
   ASSERT_TRUE(value.isNumber());
-  ASSERT_TRUE(value.toInt64() == 1);
+  EXPECT_EQ(2, value.toInt64());
+  value = block->getValue(1, 0);
+  ASSERT_TRUE(value.isNumber());
+  EXPECT_EQ(3, value.toInt64());
+  value = block->getValue(2, 0);
+  ASSERT_TRUE(value.isNumber());
+  EXPECT_EQ(4, value.toInt64());
+}
+
+class LimitExecutorTestBase {
+ protected:
+  ResourceMonitor monitor;
+  AqlItemBlockManager itemBlockManager;
+  std::shared_ptr<const std::unordered_set<RegisterId>> outputRegisters;
+  std::shared_ptr<const std::unordered_set<RegisterId>> registersToKeep;
+
+  LimitExecutorTestBase()
+      : monitor(),
+        itemBlockManager(&monitor),
+        outputRegisters(std::make_shared<const std::unordered_set<RegisterId>>(
+            std::initializer_list<RegisterId>{})),
+        registersToKeep(std::make_shared<const std::unordered_set<RegisterId>>(
+            std::initializer_list<RegisterId>{0})) {}
+};
+
+// skip and fullCount cannot go together: Only the last limit block may get
+// fullCount, so there is no block after that could skip.
+// For these cases, use this class.
+class LimitExecutorWaitingTest : public LimitExecutorTestBase,
+                                 public ::testing::TestWithParam<bool> {
+ protected:
+  bool waiting{};
+
+  LimitExecutorWaitingTest() : LimitExecutorTestBase() {}
+
+  virtual void SetUp() { waiting = GetParam(); }
+};
+
+// Fields:
+//  [0] bool waiting
+//  [1] bool fullCount
+using ExtendedLimitTestParameters = std::tuple<bool, bool>;
+
+class LimitExecutorWaitingFullCountTest
+    : public LimitExecutorTestBase,
+      public ::testing::TestWithParam<ExtendedLimitTestParameters> {
+ protected:
+  bool waiting{};
+  bool fullCount{};
+
+  LimitExecutorWaitingFullCountTest() : LimitExecutorTestBase() {}
+
+  virtual void SetUp() {
+    ExtendedLimitTestParameters const& params = GetParam();
+    std::tie(waiting, fullCount) = params;
+  }
+};
+
+void removeWaiting(std::vector<ExecutorStepResult>& results) {
+  std::vector<ExecutorStepResult> tmp;
+  for (auto const result : results) {
+    if (std::get<ExecutionState>(result) != ExecutionState::WAITING) {
+      tmp.emplace_back(result);
+    }
+  }
+  results.swap(tmp);
 }
 
+TEST_P(LimitExecutorWaitingFullCountTest, rows_9_blocksize_3_limit_10) {
+  // Input spec:
+  size_t constexpr blocksize = 3;
+  size_t constexpr offset = 0;
+  size_t constexpr limit = 10;
+  SharedAqlItemBlockPtr const input =
+      vPackBufferToAqlItemBlock(itemBlockManager, R"=( [ [0], [1], [2], [3], [4], [5], [6], [7], [8] ] )="_vpack);
```

Comment:
It might be nicer to user `buildBlock` from `tests/Aql/AqlItemBlockHelper.h` instead of `vPackBufferToAqlItemBlock`.